src/cargo/ops/cargo_package.rs

```diff
@@ -312,27 +381,49 @@ fn build_ar_list(
         let readme_path = Path::new(readme);
         let abs_file_path = paths::normalize_path(&pkg.root().join(readme_path));
         if abs_file_path.exists() {
-            check_for_file_and_add("readme", readme_path, abs_file_path, pkg, &mut result, ws)?;
+            check_for_file_and_add("readme", readme_path, abs_file_path, pkg, &mut files, ws)?;
         }
     }
-    if let Some(build_script_path) = pkg.manifest().build_script() {
-        let abs_build_script_path = paths::normalize_path(&pkg.root().join(build_script_path));
-        if abs_build_script_path.exists() {
-            check_for_file_and_add(
-                "build-script",
-                build_script_path,
-                abs_build_script_path,
-                pkg,
-                &mut result,
-                ws,
-            )?;
+
+    // Check that every target contains a path inside the package root
+    for (tgt, src_path) in pkg.manifest().targets().iter().filter_map(|tgt| {
+        if let TargetSourcePath::Path(path) = tgt.src_path() {
+            Some((tgt, path.as_ref()))
         } else {
-            unreachable!();
+            None
+        }
+    }) {
+        // If the target has a source outside, report it as an error
+        let abs_src_path = normalize_path(src_path);
+        if is_outsider(&abs_src_path, pkg.root()) {
```

Comment:
If I were you I wouldn't extract this as a function, as it is only used in one place and pretty clear for conveying its intent.

Reply:
ok